app/src/main/java/com/example/myweatherapp/viewmodel/MainViewModel.kt

```diff
@@ -0,0 +1,157 @@
+                    }
+                }
+                progressVisibilityLiveData.postValue(false)
+            }
+        }
+    }
+
+    fun loadNewsData() {
+        progressVisibilityLiveData.postValue(true)
+        return repository.getNews("news") { list ->
+            if (list != null) {
+                newsData.postValue(list)
+                applyIsFavoriteState(list)
+                Log.d("TAG", "GETTING NEWS DATA")
+            } else {
+                Log.d("TAG", "SOME ERROR")
+                progressVisibilityLiveData.postValue(false)
+            }
+        }
+    }
+
+    fun loadSourceData() {
+        progressVisibilityLiveData.postValue(true)
+        return repository.getSource { list ->
+            if (list != null) {
+                newsData.postValue(list)
+                Log.d("TAG", "GETTING SOURCE DATA")
+                progressVisibilityLiveData.postValue(false)
+            } else {
+                Log.d("TAG", "Some ERROR")
+            }
+        }
+    }
+
+    fun loadFavoriteData() {
+        progressVisibilityLiveData.postValue(true)
+        return repository.getNews("news") { list ->
+            if (list != null) {
+                val favoriteKeyUrls = sharedPreferences.getStringSet(KEY_URL, emptySet()) ?: emptySet()
+                val filteredData = list.filter { item ->
+                    favoriteKeyUrls.any {
+                        item.newsUrl == it
+                    }
+                }
+                applyIsFavoriteState(list)
+                favoriteData.postValue(filteredData)
+                Log.d("TAG", "GETTING NEWS DATA")
+                progressVisibilityLiveData.postValue(false)
+            } else {
+                Log.d("TAG", "SOME ERROR")
+            }
+        }
+    }
+
+    fun applyIsFavoriteState(list: List<NewsModel>) {
+        val favoriteState = sharedPreferences.getStringSet(KEY_URL, emptySet()) ?: emptySet()
+        list.forEach { item ->
+            item.isFavorite =
+                favoriteState.any { url ->
+                    url == item.newsUrl
+                }
+        }
+    }
+
+    internal fun addNewsToFavoriteList(url: String) {
+        val prefsKeys = sharedPreferences.getStringSet(KEY_URL, emptySet())
+        val mutableKeySet = mutableSetOf<String>()
+
+        sharedPreferences.edit()?.let { editor ->
+            if (prefsKeys.isNullOrEmpty()) {
+                editor.putStringSet(KEY_URL, setOf(url))
+            } else {
+                mutableKeySet.apply {
+                    addAll(prefsKeys)
+                    add(url)
+                    Log.d("TAG", "ADDING NEWS TO FAVORITE LIST")
+                }
+                editor.putStringSet(KEY_URL, mutableKeySet)
+            }
+        }?.apply()
+    }
+
+    internal fun deleteNewsFromFavoriteList(url: String) {
+        val prefsKeys = sharedPreferences.getStringSet(KEY_URL, emptySet()) ?: emptySet()
```

Comment:
you do use that `sharedPreferences.getStringSet(KEY_URL, emptySet())` many times in the class. Can't you extract that into a method and reuse it?

Reply:
added method to get KEY_URLs